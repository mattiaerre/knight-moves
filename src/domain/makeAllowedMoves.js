import { files, ranks } from './constants';

// [file, rank]

function up(position) {
  const allowedMoves = [];
  allowedMoves.push([files[files.indexOf(position[0]) - 1], position[1] + 2]);
  allowedMoves.push([files[files.indexOf(position[0]) + 1], position[1] + 2]);
  return allowedMoves;
}

function right(position) {
  const allowedMoves = [];
  allowedMoves.push([files[files.indexOf(position[0]) + 2], position[1] + 1]);
  allowedMoves.push([files[files.indexOf(position[0]) + 2], position[1] - 1]);
  return allowedMoves;
}

function down(position) {
  const allowedMoves = [];
  allowedMoves.push([files[files.indexOf(position[0]) + 1], position[1] - 2]);
  allowedMoves.push([files[files.indexOf(position[0]) - 1], position[1] - 2]);
  return allowedMoves;
}

function left(position) {
  const allowedMoves = [];
  allowedMoves.push([files[files.indexOf(position[0]) - 2], position[1] - 1]);
  allowedMoves.push([files[files.indexOf(position[0]) - 2], position[1] + 1]);
  return allowedMoves;
}

function makeAllowedMoves(position) {
  // INFO: moving clockwise from "position"
  return [up(position)[1]]
    .concat(right(position))
    .concat(down(position))
    .concat(left(position))
    .concat([up(position)[0]])
    .filter(([file, rank]) => files.includes(file) && ranks.includes(rank));
}

export default makeAllowedMoves;
