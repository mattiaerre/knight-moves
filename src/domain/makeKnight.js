import makeAllowedMoves from './makeAllowedMoves';

function makeKnight(initial) {
  function canMoveTo(destination) {
    return (
      makeAllowedMoves(initial).find(
        ([file, rank]) => file === destination[0] && rank === destination[1]
      ) !== undefined
    );
  }

  return {
    canMoveTo
  };
}

export default makeKnight;
