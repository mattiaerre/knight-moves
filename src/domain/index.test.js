import makeAllowedMoves from './makeAllowedMoves';
import makeKnight from './makeKnight';

describe('allowed moves', () => {
  test('b1', () => {
    const allowedMoves = makeAllowedMoves(['b', 1]);
    expect(JSON.stringify(allowedMoves)).toEqual(
      JSON.stringify([
        ['c', 3],
        ['d', 2],
        ['a', 3]
      ])
    );
  });

  test('e5', () => {
    const allowedMoves = makeAllowedMoves(['e', 5]);
    expect(JSON.stringify(allowedMoves)).toEqual(
      JSON.stringify([
        ['f', 7],
        ['g', 6],
        ['g', 4],
        ['f', 3],
        ['d', 3],
        ['c', 4],
        ['c', 6],
        ['d', 7]
      ])
    );
  });

  test('h6', () => {
    const allowedMoves = makeAllowedMoves(['h', 6]);
    expect(allowedMoves).toMatchSnapshot();
  });
});

describe('knight', () => {
  describe('can move to', () => {
    it('c3', () => {
      const knight = makeKnight(['b', 1]);
      expect(knight.canMoveTo(['c', 3])).toBe(true);
    });

    it('b3', () => {
      const knight = makeKnight(['b', 1]);
      expect(knight.canMoveTo(['b', 3])).toBe(false);
    });

    it('a3', () => {
      const knight = makeKnight(['b', 1]);
      expect(knight.canMoveTo(['a', 3])).toBe(true);
    });

    it('h8', () => {
      const knight = makeKnight(['h', 8]);
      expect(knight.canMoveTo(['g', 6])).toBe(true);
    });
  });

  describe('move to', () => {
    it('b1 > c3 > d5', () => {
      const knight = makeKnight(['b', 1]);
      if (knight.canMoveTo(['c', 3])) {
        knight.moveTo(['c', 3]);
      }
      if (knight.canMoveTo(['d', 5])) {
        knight.moveTo(['d', 5]);
      }
      expect(knight.getPosition()).toEqual(['d', 5]);
    });
  });
});
