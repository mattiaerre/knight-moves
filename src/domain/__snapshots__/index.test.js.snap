// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allowed moves h6 1`] = `
Array [
  Array [
    "g",
    4,
  ],
  Array [
    "f",
    5,
  ],
  Array [
    "f",
    7,
  ],
  Array [
    "g",
    8,
  ],
]
`;
